### ISSUE TITLE
canvas not taken as reference

### ISSUE THREAD

**Issue by Kethku** [rust-skia/rust-skia#398]
https://github.com/rust-skia/rust-skia/blob/f4709ab7336a25b3e411c8c3e2ea4331eb3ff2a8/skia-safe/src/core/surface.rs#L380

This line consumes the canvas passed to it. I think that may be incorrect?

**Comment by Kethku** [comment]
I'm an idiot. This is my problem

**Comment by Kethku** [comment]
Actually yes. I believe this is a problem. The canvas should be &mut not just mut.

**Comment by pragmatrix** [comment]
The function should accept both, a canvas by taking ownership, and canvas - the most likely case - by a mut reference. At least that was the original intention, which - I'll admit - might be confusing. 

``` rust
#[test]
fn test_drawing_as_mut_ergonomics() {
    let mut surface = Surface::new_raster_n32_premul((16, 16)).unwrap();

    // option1:
    // - An _owned_ canvas can be drawn by transferring ownership.
    {
        let canvas = Canvas::new(ISize::new(16, 16), None).unwrap();
        surface.draw(canvas, (10.0, 10.0), None);
    }

    // option2:
    // - An &mut canvas can be drawn.

    {
        let mut canvas = Canvas::new(ISize::new(16, 16), None).unwrap();
        surface.draw(&mut canvas, (10.0, 10.0), None);
    }

    // option3:
    // - A canvas from another surface can be drawn.
    {
        let mut surface2 = Surface::new_raster_n32_premul((16, 16)).unwrap();
        let canvas = surface2.canvas();
        surface.draw(canvas, (10.0, 10.0), None);
    }
}
```

See also: https://github.com/rust-skia/rust-skia/blob/94cd93e9e1aa8f6ddca8ae724809a286521c8a6b/skia-safe/src/core/canvas.rs#L180

And I am absolutely not sure if this is in any way idiomatic Rust.

**Comment by Kethku** [comment]
https://github.com/rust-skia/rust-skia/pull/399/files#r481846944 this gives an example closer to my problem. I'm very much not a Rust expert though so there is very possibly a gap in my understanding. Thanks so much for humoring me thus far.

**Comment by Kethku** [comment]
I see your second point about AsMut. So I guess the fix for my issue is to just call as_mut manually rather than passing the mutable reference?

**Comment by pragmatrix** [comment]
Closed via #402. Thank you for being persistent here. I now think that the removal of the `AsMut` trait from the function signatures that expect a mutable Canvas reference only, is beneficial and less confusing for most uses.
